File: python/06-nondimensionalization.py
"""
Physics Through Computational Thinking
Section 1.6: Non-Dimensionalization - Making Physics Universal

This script demonstrates non-dimensionalization techniques for
transforming physics problems into universal mathematical form.

Requirements:
    pip install numpy plotly
"""

import numpy as np
import plotly.graph_objects as go
from plotly.subplots import make_subplots


# ============================================================
# Example 1: Pendulum Energy (Dimensional vs Dimensionless)
# ============================================================

def example_pendulum_energy():
    """Compare dimensional and dimensionless pendulum potential."""
    print("\n=== Example 1: Pendulum Energy ===")

    # Physical parameters (these will vary)
    m = 2.0   # mass (kg)
    g = 9.8   # gravity (m/s²)
    l = 1.5   # length (m)

    # Natural energy scale
    E_scale = m * g * l
    print(f"Energy scale mgℓ = {E_scale:.2f} J")

    # Angle array
    theta = np.linspace(-np.pi, np.pi, 200)
    theta_deg = theta * 180 / np.pi

    # Dimensional potential (Joules)
    V_dim = m * g * l * (1 - np.cos(theta))

    # Dimensionless potential (universal!)
    V_bar = 1 - np.cos(theta)

    # Create side-by-side comparison
    fig = make_subplots(
        rows=1, cols=2,
        subplot_titles=[
            f'Dimensional V(θ) [m={m}kg, g={g}m/s², ℓ={l}m]',
            'Dimensionless V̄(θ) = 1 - cos θ'
        ]
    )

    fig.add_trace(go.Scatter(
        x=theta_deg, y=V_dim,
        mode='lines', line=dict(color='#00f3ff', width=2),
        name='V (Joules)'
    ), row=1, col=1)

    fig.add_trace(go.Scatter(
        x=theta_deg, y=V_bar,
        mode='lines', line=dict(color='#ff00ff', width=2),
        name='V̄ (dimensionless)'
    ), row=1, col=2)

    fig.update_layout(
        template='plotly_dark',
        title='Pendulum Potential Energy: Dimensional vs Dimensionless',
        showlegend=False
    )

    fig.update_xaxes(title_text='θ (degrees)', gridcolor='#2a2f4a')
    fig.update_yaxes(gridcolor='#2a2f4a', zerolinecolor='#808080')

    fig.show()


# ============================================================
# Example 2: Projectile on Hill
# ============================================================

def projectile_on_hill(theta_deg, phi_deg, Gamma, n_points=100):
    """
    Compute dimensionless trajectory and find landing point.

    Parameters:
        theta_deg: Launch angle (degrees)
        phi_deg: Hill slope angle (degrees)
        Gamma: Dimensionless parameter gh/v²
        n_points: Number of trajectory points

    Returns:
        Dictionary with trajectory and landing data
    """
    theta = np.radians(theta_deg)
    phi = np.radians(phi_deg)

    # Trajectory: y = -Γx²/(2cos²θ) + x·tanθ + Γ
    # Hill: y = -x·tanφ + Γ
    # Intersection: solve for x

    A = -Gamma / (2 * np.cos(theta)**2)
    B = np.tan(theta) + np.tan(phi)

    if abs(A) < 1e-10:
        x_land = 0
    else:
        x_land = -B / A

    x_land = max(0, x_land)
    y_land = -np.tan(phi) * x_land + Gamma

    # Generate trajectory
    x_max = max(x_land * 1.2, 2)
    x_traj = np.linspace(0, x_max, n_points)
    y_traj = A * x_traj**2 + np.tan(theta) * x_traj + Gamma

    # Hill line
    x_hill = np.linspace(0, x_max * 1.1, 50)
    y_hill = -np.tan(phi) * x_hill + Gamma

    return {
        'x_traj': x_traj, 'y_traj': y_traj,
        'x_hill': x_hill, 'y_hill': y_hill,
        'x_land': x_land, 'y_land': y_land,
        'x_max': x_max
    }


def example_projectile_on_hill():
    """Demonstrate projectile landing on sloped terrain."""
    print("\n=== Example 2: Projectile on Hill ===")

    # Parameters
    theta, phi, Gamma = 45, 20, 0.3
    print(f"Launch angle θ = {theta}°")
    print(f"Hill slope φ = {phi}°")
    print(f"Γ = gh/v² = {Gamma}")

    result = projectile_on_hill(theta, phi, Gamma)

    fig = go.Figure()

    # Trajectory
    fig.add_trace(go.Scatter(
        x=result['x_traj'], y=result['y_traj'],
        mode='lines', line=dict(color='#00f3ff', width=2),
        name='Trajectory'
    ))

    # Hill
    fig.add_trace(go.Scatter(
        x=result['x_hill'], y=result['y_hill'],
        mode='lines', line=dict(color='#ffff00', width=2),
        name='Hill'
    ))

    # Start point
    fig.add_trace(go.Scatter(
        x=[0], y=[Gamma],
        mode='markers', marker=dict(color='#00ff9f', size=10),
        name='Launch'
    ))

    # Landing point
    fig.add_trace(go.Scatter(
        x=[result['x_land']], y=[result['y_land']],
        mode='markers', marker=dict(color='#ff00ff', size=12),
        name='Landing'
    ))

    fig.update_layout(
        template='plotly_dark',
        title=f'Projectile on Hill: θ={theta}°, φ={phi}°, Γ={Gamma}',
        xaxis=dict(
            title='x̄ (dimensionless)',
            gridcolor='#2a2f4a',
            zerolinecolor='#808080',
            scaleanchor='y',
            range=[-0.3, result['x_max'] * 1.1]
        ),
        yaxis=dict(
            title='ȳ (dimensionless)',
            gridcolor='#2a2f4a',
            zerolinecolor='#808080',
            range=[-0.5, Gamma * 1.5]
        )
    )

    fig.show()

    print(f"Landing point: x̄ = {result['x_land']:.3f}, ȳ = {result['y_land']:.3f}")


# ============================================================
# Example 3: Finding Optimal Launch Angle
# ============================================================

def example_optimal_angle():
    """Find optimal launch angle for maximum range."""
    print("\n=== Example 3: Optimal Launch Angle ===")

    phi = 20  # Hill slope
    Gamma = 0.3  # Dimensionless height

    angles = np.linspace(10, 80, 50)
    ranges = []

    for theta in angles:
        result = projectile_on_hill(theta, phi, Gamma)
        ranges.append(result['x_land'])

    ranges = np.array(ranges)
    optimal_idx = np.argmax(ranges)
    optimal_angle = angles[optimal_idx]
    max_range = ranges[optimal_idx]

    fig = go.Figure()

    fig.add_trace(go.Scatter(
        x=angles, y=ranges,
        mode='lines', line=dict(color='#00f3ff', width=2),
        name='Range'
    ))

    fig.add_trace(go.Scatter(
        x=[optimal_angle], y=[max_range],
        mode='markers', marker=dict(color='#ff00ff', size=12),
        name=f'Optimal: {optimal_angle:.1f}°'
    ))

    fig.update_layout(
        template='plotly_dark',
        title=f'Range vs Launch Angle (φ={phi}°, Γ={Gamma})',
        xaxis=dict(title='Launch angle θ (degrees)', gridcolor='#2a2f4a'),
        yaxis=dict(title='Dimensionless range x̄', gridcolor='#2a2f4a')
    )

    fig.show()

    print(f"Optimal launch angle: {optimal_angle:.1f}°")
    print(f"Maximum range: {max_range:.3f}")


# ============================================================
# Example 4: Damped Harmonic Oscillator
# ============================================================

def example_damped_oscillator():
    """Non-dimensionalize the damped harmonic oscillator."""
    print("\n=== Example 4: Damped Harmonic Oscillator ===")

    # Original equation: m*d²x/dt² = -k*x - b*dx/dt
    # Variables: m, k, b, x, t (5 variables, 3 dimensions MLT)
    # → 2 dimensionless parameters!

    print("Original equation: m·ẍ = -kx - b·ẋ")
    print("\nNatural scales:")
    print("  Time scale: 1/ω₀ = √(m/k)")
    print("  Length scale: x₀ (initial displacement)")
    print("\nDimensionless variables:")
    print("  τ = t·ω₀ = t·√(k/m)")
    print("  X = x/x₀")
    print("  ζ = b/(2√(km))  [damping ratio]")
    print("\nDimensionless equation: d²X/dτ² = -X - 2ζ·dX/dτ")

    # Solve for different damping ratios
    tau = np.linspace(0, 20, 500)

    fig = go.Figure()

    colors = ['#00f3ff', '#ff00ff', '#ffff00', '#00ff9f']
    zetas = [0.1, 0.3, 0.7, 1.0]

    for zeta, color in zip(zetas, colors):
        if zeta < 1:  # Underdamped
            omega_d = np.sqrt(1 - zeta**2)
            X = np.exp(-zeta * tau) * np.cos(omega_d * tau)
            label = f'ζ = {zeta} (underdamped)'
        elif zeta == 1:  # Critically damped
            X = (1 + tau) * np.exp(-tau)
            label = f'ζ = {zeta} (critical)'
        else:  # Overdamped
            r1 = -zeta + np.sqrt(zeta**2 - 1)
            r2 = -zeta - np.sqrt(zeta**2 - 1)
            X = 0.5 * (np.exp(r1 * tau) + np.exp(r2 * tau))
            label = f'ζ = {zeta} (overdamped)'

        fig.add_trace(go.Scatter(
            x=tau, y=X,
            mode='lines', line=dict(color=color, width=2),
            name=label
        ))

    fig.update_layout(
        template='plotly_dark',
        title='Damped Oscillator: Universal Dimensionless Solution',
        xaxis=dict(
            title='τ = t·√(k/m)',
            gridcolor='#2a2f4a',
            zerolinecolor='#808080'
        ),
        yaxis=dict(
            title='X = x/x₀',
            gridcolor='#2a2f4a',
            zerolinecolor='#808080',
            range=[-1.1, 1.1]
        ),
        legend=dict(x=0.7, y=0.95)
    )

    fig.show()


# ============================================================
# Example 5: Scale Identification Practice
# ============================================================

def example_scale_identification():
    """Practice identifying natural scales for various problems."""
    print("\n=== Example 5: Scale Identification ===")

    problems = {
        'Pendulum': {
            'equation': 'θ̈ = -(g/ℓ)·sin θ',
            'variables': ['m', 'g', 'ℓ', 'θ'],
            'scales': {
                'Length': 'ℓ',
                'Time': '√(ℓ/g)',
                'Energy': 'mgℓ'
            }
        },
        'Spring-Mass': {
            'equation': 'mẍ = -kx',
            'variables': ['m', 'k', 'x'],
            'scales': {
                'Time': '√(m/k)',
                'Length': 'x₀',
                'Energy': '½kx₀²'
            }
        },
        'Projectile': {
            'equation': 'ÿ = -g',
            'variables': ['v', 'g', 'θ'],
            'scales': {
                'Length': 'v²/g',
                'Time': 'v/g',
                'Velocity': 'v'
            }
        },
        'Falling with drag': {
            'equation': 'mv̇ = mg - bv',
            'variables': ['m', 'g', 'b'],
            'scales': {
                'Velocity': 'mg/b (terminal)',
                'Time': 'm/b',
                'Length': 'm²g/b²'
            }
        }
    }

    for name, data in problems.items():
        print(f"\n{name}:")
        print(f"  Equation: {data['equation']}")
        print(f"  Variables: {', '.join(data['variables'])}")
        print("  Natural scales:")
        for qty, scale in data['scales'].items():
            print(f"    {qty}: {scale}")


# ============================================================
# Example 6: Buckingham Pi Theorem
# ============================================================

def example_buckingham_pi():
    """Demonstrate the Buckingham Pi theorem."""
    print("\n=== Example 6: Buckingham Pi Theorem ===")

    print("The Buckingham Pi theorem:")
    print("  If a problem has N variables and M fundamental dimensions,")
    print("  then there are (N - M) independent dimensionless groups.")
    print()

    examples = [
        {
            'name': 'Pendulum period',
            'variables': ['T', 'm', 'g', 'ℓ'],
            'N': 4,
            'dimensions': ['M', 'L', 'T'],
            'M': 3,
            'result': 'T/√(ℓ/g) = f(nothing) = constant = 2π'
        },
        {
            'name': 'Projectile range',
            'variables': ['R', 'v', 'g', 'θ'],
            'N': 4,
            'dimensions': ['L', 'T'],
            'M': 2,
            'result': 'Rg/v² = f(θ) → R = (v²/g)·sin(2θ)'
        },
        {
            'name': 'Drag force',
            'variables': ['F', 'ρ', 'v', 'D'],
            'N': 4,
            'dimensions': ['M', 'L', 'T'],
            'M': 3,
            'result': 'F/(ρv²D²) = C_D (drag coefficient)'
        }
    ]

    for ex in examples:
        print(f"{ex['name']}:")
        print(f"  Variables: {', '.join(ex['variables'])} (N = {ex['N']})")
        print(f"  Dimensions: {', '.join(ex['dimensions'])} (M = {ex['M']})")
        print(f"  Dimensionless groups: N - M = {ex['N'] - ex['M']}")
        print(f"  Result: {ex['result']}")
        print()


# ============================================================
# Main
# ============================================================

if __name__ == "__main__":
    print("=" * 60)
    print("Section 1.6: Non-Dimensionalization - Making Physics Universal")
    print("=" * 60)
    print("\nThis script demonstrates:")
    print("1. Pendulum: dimensional vs dimensionless energy")
    print("2. Projectile on a hill with dimensionless parameters")
    print("3. Finding optimal launch angles")
    print("4. Damped harmonic oscillator")
    print("5. Scale identification for various problems")
    print("6. Buckingham Pi theorem")

    example_pendulum_energy()
    example_projectile_on_hill()
    example_optimal_angle()
    example_damped_oscillator()
    example_scale_identification()
    example_buckingham_pi()

    print("\n" + "=" * 60)
    print("All examples complete!")
    print("=" * 60)

File: js/sections/06-nondimensionalization.js
/**
 * Section 1.6: Non-Dimensionalization - Making Physics Universal
 *
 * Interactive explorers:
 * 1. Pendulum Energy (dimensional vs dimensionless comparison)
 * 2. Projectile on Hill (3 dimensionless parameters)
 * 3. Scale Identifier (practice finding natural scales)
 */

(function() {
    'use strict';

    // ========================================
    // Initialization
    // ========================================

    function initializePlots() {
        // Check dependencies
        if (typeof Plotly === 'undefined') {
            setTimeout(initializePlots, 100);
            return;
        }

        // Check DOM elements exist
        const pendulumDim = document.getElementById('nondim-pendulum-dimensional');
        const pendulumNondim = document.getElementById('nondim-pendulum-dimensionless');
        const projectilePlot = document.getElementById('nondim-projectile-plot');
        const scalesPlot = document.getElementById('nondim-scales-plot');

        if (!pendulumDim || !pendulumNondim || !projectilePlot || !scalesPlot) {
            setTimeout(initializePlots, 100);
            return;
        }

        // Initialize all explorers
        initPendulumExplorer();
        initProjectileExplorer();
        initScalesExplorer();
        attachEventHandlers();
    }

    // ========================================
    // Explorer 1: Pendulum Energy
    // ========================================

    function initPendulumExplorer() {
        updatePendulumPlots();
    }

    function updatePendulumPlots() {
        const m = parseFloat(document.getElementById('nondim-mass-slider').value);
        const l = parseFloat(document.getElementById('nondim-length-slider').value);
        const g = parseFloat(document.getElementById('nondim-gravity-slider').value);

        // Update displayed values
        document.getElementById('nondim-mass-value').textContent = m.toFixed(1);
        document.getElementById('nondim-length-value').textContent = l.toFixed(1);
        document.getElementById('nondim-gravity-value').textContent = g.toFixed(1);

        // Energy scale (displayed as the dimensional plot's title)
        const E_scale = m * g * l;

        // Generate data
        const theta = [];
        const V_dim = [];
        const V_nondim = [];
        const n = 200;

        for (let i = 0; i < n; i++) {
            const t = -Math.PI + (2 * Math.PI * i / (n - 1));
            theta.push(t * 180 / Math.PI); // Convert to degrees for display
            V_dim.push(m * g * l * (1 - Math.cos(t)));
            V_nondim.push(1 - Math.cos(t));
        }

        // Dimensional plot
        const traceDim = {
            x: theta,
            y: V_dim,
            type: 'scatter',
            mode: 'lines',
            line: { color: '#00f3ff', width: 2.5 },
            name: 'V(θ)'
        };

        const layoutDim = {
            template: 'plotly_dark',
            paper_bgcolor: '#0a0e27',
            plot_bgcolor: '#151934',
            margin: { l: 60, r: 30, t: 30, b: 50 },
            title: {
                text: 'energy scale mgℓ = ' + E_scale.toFixed(1) + ' J',
                font: { size: 12, color: '#a0a8d0' }
            },
            xaxis: {
                title: 'θ (degrees)',
                range: [-180, 180],
                gridcolor: '#2a2f4a',
                zerolinecolor: '#808080'
            },
            yaxis: {
                title: 'V (Joules)',
                // FIXED range: if this axis rescaled with mgℓ the dimensional
                // curve would always fill the same fraction of the plot and
                // look identical to the dimensionless one — hiding the whole
                // point. With fixed axes, changing m, ℓ, or g visibly grows or
                // shrinks V(θ) while the dimensionless panel never moves.
                range: [0, 100],
                gridcolor: '#2a2f4a',
                zerolinecolor: '#808080'
            },
            showlegend: false
        };

        Plotly.react('nondim-pendulum-dimensional', [traceDim], layoutDim, { responsive: true });

        // Dimensionless plot
        const traceNondim = {
            x: theta,
            y: V_nondim,
            type: 'scatter',
            mode: 'lines',
            line: { color: '#ff00ff', width: 2.5 },
            name: 'V̄(θ)'
        };

        const layoutNondim = {
            template: 'plotly_dark',
            paper_bgcolor: '#0a0e27',
            plot_bgcolor: '#151934',
            margin: { l: 60, r: 30, t: 30, b: 50 },
            xaxis: {
                title: 'θ (degrees)',
                range: [-180, 180],
                gridcolor: '#2a2f4a',
                zerolinecolor: '#808080'
            },
            yaxis: {
                title: 'V̄ = V/(mgℓ)',
                range: [0, 2.2],
                gridcolor: '#2a2f4a',
                zerolinecolor: '#808080'
            },
            showlegend: false
        };

        Plotly.react('nondim-pendulum-dimensionless', [traceNondim], layoutNondim, { responsive: true });
    }

    // ========================================
    // Explorer 2: Projectile on Hill
    // ========================================

    function initProjectileExplorer() {
        updateProjectilePlot();
    }

    function updateProjectilePlot() {
        const thetaDeg = parseFloat(document.getElementById('nondim-theta-slider').value);
        const phiDeg = parseFloat(document.getElementById('nondim-phi-slider').value);
        const Gamma = parseFloat(document.getElementById('nondim-gamma-slider').value);

        // Update displayed values
        document.getElementById('nondim-theta-value').textContent = thetaDeg;
        document.getElementById('nondim-phi-value').textContent = phiDeg;
        document.getElementById('nondim-gamma-value').textContent = Gamma.toFixed(2);

        // Convert to radians
        const theta = thetaDeg * Math.PI / 180;
        const phi = phiDeg * Math.PI / 180;

        // Non-dimensionalized with the HILL HEIGHT h as the length scale
        // (x̄ = x/h, ȳ = y/h), so the terrain is anchored: launch always at
        // (0, 1), hill base always at (cot φ, 0), flat ground ȳ = 0 beyond.
        // Γ = gh/v² enters only through the trajectory's curvature:
        //   Trajectory: ȳ = 1 + x̄·tanθ − Γ x̄²/(2cos²θ)
        //   Hill face:  ȳ = 1 − x̄·tanφ   (0 ≤ x̄ ≤ cot φ), then ground ȳ = 0
        const tanT = Math.tan(theta);
        const tanP = Math.tan(phi);
        const cos2 = Math.cos(theta) * Math.cos(theta);
        const xBase = tanP > 1e-10 ? 1 / tanP : Infinity;

        // Landing: try the hill face first, else the flat ground
        let xLand = 2 * cos2 * (tanT + tanP) / Gamma;
        let yLand, regime;
        if (xLand <= xBase) {
            yLand = 1 - xLand * tanP;
            regime = 'hill face';
        } else {
            xLand = (cos2 / Gamma) * (tanT + Math.sqrt(tanT * tanT + 2 * Gamma / cos2));
            yLand = 0;
            regime = 'flat ground';
        }

        // Fixed axes: with the terrain anchored and Γ ≥ 0.2 every
        // trajectory (apex ≤ 1 + 1/(2Γ) ≤ 3.5, landing ≤ ~6) stays in frame
        const xAxisMax = 6.5;
        const yAxisMax = 3.6;

        // Trajectory drawn from launch to the landing point
        const xTraj = [];
        const yTraj = [];
        const nPoints = 120;
        for (let i = 0; i < nPoints; i++) {
            const x = xLand * i / (nPoints - 1);
            yTraj.push(1 + tanT * x - Gamma * x * x / (2 * cos2));
            xTraj.push(x);
        }

        // Terrain polyline: hill face down to its base, then flat ground —
        // one continuous yellow line
        const xTerr = [0];
        const yTerr = [1];
        if (xBase < xAxisMax) {
            xTerr.push(xBase, xAxisMax);
            yTerr.push(0, 0);
        } else {
            xTerr.push(xAxisMax);
            yTerr.push(1 - xAxisMax * tanP);
        }

        const traceTrajectory = {
            x: xTraj,
            y: yTraj,
            type: 'scatter',
            mode: 'lines',
            line: { color: '#00f3ff', width: 2.5 },
            name: 'Trajectory'
        };

        const traceTerrain = {
            x: xTerr,
            y: yTerr,
            type: 'scatter',
            mode: 'lines',
            line: { color: '#ffff00', width: 2 },
            name: 'Hill & ground'
        };

        const traceLanding = {
            x: [xLand],
            y: [yLand],
            type: 'scatter',
            mode: 'markers',
            marker: { color: '#ff00ff', size: 12, symbol: 'circle' },
            name: 'Landing'
        };

        // Launch is anchored at (0, 1) = the hilltop, one hill-height up
        const traceStart = {
            x: [0],
            y: [1],
            type: 'scatter',
            mode: 'markers',
            marker: { color: '#00ff9f', size: 10, symbol: 'circle' },
            name: 'Launch'
        };

        const layout = {
            template: 'plotly_dark',
            paper_bgcolor: '#0a0e27',
            plot_bgcolor: '#151934',
            margin: { l: 60, r: 30, t: 30, b: 50 },
            xaxis: {
                title: 'x̄ = x/h',
                range: [-0.2, xAxisMax],
                gridcolor: '#2a2f4a',
                zerolinecolor: '#808080'
            },
            yaxis: {
                title: 'ȳ = y/h',
                range: [-0.15, yAxisMax],
                gridcolor: '#2a2f4a',
                zerolinecolor: '#808080'
            },
            showlegend: true,
            legend: {
                x: 0.98,
                xanchor: 'right',
                y: 0.98,
                bgcolor: 'rgba(10, 14, 39, 0.8)'
            }
        };

        Plotly.react('nondim-projectile-plot',
            [traceTrajectory, traceTerrain, traceStart, traceLanding],
            layout,
            { responsive: true }
        );

        const statsDiv = document.getElementById('nondim-projectile-stats');
        if (statsDiv) {
            statsDiv.innerHTML =
                '<span>landing x̄ = <strong>' + xLand.toFixed(3) + '</strong></span>' +
                '<span>lands on: <strong>' + regime + '</strong></span>' +
                '<span>launch speed V = v/√(gh) = 1/√Γ = <strong>' + (1 / Math.sqrt(Gamma)).toFixed(2) + '</strong></span>';
        }
    }

    // ========================================
    // Explorer 3: Scale Identifier
    // ========================================

    const scaleProblems = {
        pendulum: {
            title: 'Simple Pendulum',
            equation: 'θ̈ = -(g/ℓ) sin θ',
            variables: ['m (mass)', 'g (gravity)', 'ℓ (length)', 'θ (angle)'],
            scales: {
                'Length': 'ℓ',
                'Time': '√(ℓ/g)',
                'Energy': 'mgℓ',
                'Angular freq.': '√(g/ℓ)'
            },
            dimensionless: 'θ̈ = -sin θ  (using τ = t√(g/ℓ))',
            plotData: function() {
                const tau = [];
                const thetaData = [];
                for (let i = 0; i < 200; i++) {
                    const t = 4 * Math.PI * i / 199;
                    tau.push(t);
                    thetaData.push(0.3 * Math.cos(t)); // small angle approx
                }
                return { x: tau, y: thetaData, xlabel: 'τ = t√(g/ℓ)', ylabel: 'θ (rad)' };
            }
        },
        spring: {
            title: 'Spring-Mass Oscillator',
            equation: 'mẍ = -kx',
            variables: ['m (mass)', 'k (spring constant)', 'x (displacement)'],
            scales: {
                'Length': 'x₀ (initial displacement)',
                'Time': '√(m/k)',
                'Energy': '½kx₀²',
                'Angular freq.': '√(k/m)'
            },
            dimensionless: 'Ẍ = -X  (using τ = t√(k/m), X = x/x₀)',
            plotData: function() {
                const tau = [];
                const X = [];
                for (let i = 0; i < 200; i++) {
                    const t = 4 * Math.PI * i / 199;
                    tau.push(t);
                    X.push(Math.cos(t));
                }
                return { x: tau, y: X, xlabel: 'τ = t√(k/m)', ylabel: 'X = x/x₀' };
            }
        },
        projectile: {
            title: 'Projectile Motion',
            equation: 'ẍ = 0, ÿ = -g',
            variables: ['v (initial speed)', 'g (gravity)', 'θ (launch angle)'],
            scales: {
                'Length': 'v²/g',
                'Time': 'v/g',
                'Velocity': 'v'
            },
            dimensionless: 'x̄″ = 0, ȳ″ = -1  (using τ = tg/v, x̄ = x/(v²/g))',
            plotData: function() {
                const x = [];
                const y = [];
                const theta = 45 * Math.PI / 180;
                for (let i = 0; i < 100; i++) {
                    const tau = 2 * Math.sin(theta) * i / 99;
                    x.push(tau * Math.cos(theta));
                    y.push(tau * Math.sin(theta) - 0.5 * tau * tau);
                }
                return { x: x, y: y, xlabel: 'x̄ = x/(v²/g)', ylabel: 'ȳ = y/(v²/g)' };
            }
        },
        falling: {
            title: 'Falling with Linear Drag',
            equation: 'mv̇ = mg - bv',
            variables: ['m (mass)', 'g (gravity)', 'b (drag coefficient)'],
            scales: {
                'Velocity': 'v_term = mg/b',
                'Time': 'm/b',
                'Length': 'm²g/b² = v_term · (m/b)'
            },
            dimensionless: 'V̇ = 1 - V  (using τ = bt/m, V = v/v_term)',
            plotData: function() {
                const tau = [];
                const V = [];
                for (let i = 0; i < 200; i++) {
                    const t = 5 * i / 199;
                    tau.push(t);
                    V.push(1 - Math.exp(-t)); // solution: V = 1 - e^(-τ)
                }
                return { x: tau, y: V, xlabel: 'τ = bt/m', ylabel: 'V = v/v_term' };
            }
        }
    };

    function initScalesExplorer() {
        updateScalesDisplay();
    }

    function updateScalesDisplay() {
        const problemKey = document.getElementById('nondim-problem-select').value;
        const problem = scaleProblems[problemKey];

        // Update info display
        const display = document.getElementById('nondim-scales-display');
        let html = `
            <div class="info-box-title">${problem.title}</div>
            <p><strong>Equation:</strong> ${problem.equation}</p>
            <p><strong>Variables:</strong> ${problem.variables.join(', ')}</p>
            <table style="width: 100%; margin: var(--spacing-sm) 0;">
                <tr style="border-bottom: 1px solid var(--border-color);">
                    <th style="text-align: left; padding: 4px;">Quantity</th>
                    <th style="text-align: left; padding: 4px;">Natural Scale</th>
                </tr>
        `;
        for (const [qty, scale] of Object.entries(problem.scales)) {
            html += `<tr><td style="padding: 4px;">${qty}</td><td style="padding: 4px; color: var(--neon-cyan);">${scale}</td></tr>`;
        }
        html += `</table>
            <p><strong>Dimensionless form:</strong> <code style="color: var(--neon-magenta);">${problem.dimensionless}</code></p>
        `;
        display.innerHTML = html;

        // Update plot
        const plotData = problem.plotData();

        const trace = {
            x: plotData.x,
            y: plotData.y,
            type: 'scatter',
            mode: 'lines',
            line: { color: '#00f3ff', width: 2.5 }
        };

        const layout = {
            template: 'plotly_dark',
            paper_bgcolor: '#0a0e27',
            plot_bgcolor: '#151934',
            margin: { l: 70, r: 30, t: 30, b: 50 },
            xaxis: {
                title: plotData.xlabel,
                gridcolor: '#2a2f4a',
                zerolinecolor: '#808080'
            },
            yaxis: {
                title: plotData.ylabel,
                gridcolor: '#2a2f4a',
                zerolinecolor: '#808080'
            },
            showlegend: false
        };

        Plotly.react('nondim-scales-plot', [trace], layout, { responsive: true });
    }

    // ========================================
    // Event Handlers
    // ========================================

    function attachEventHandlers() {
        // Pendulum sliders
        ['mass', 'length', 'gravity'].forEach(param => {
            const slider = document.getElementById(`nondim-${param}-slider`);
            if (slider) {
                slider.addEventListener('input', updatePendulumPlots);
            }
        });

        // Projectile sliders
        ['theta', 'phi', 'gamma'].forEach(param => {
            const slider = document.getElementById(`nondim-${param}-slider`);
            if (slider) {
                slider.addEventListener('input', updateProjectilePlot);
            }
        });

        // Scales dropdown
        const problemSelect = document.getElementById('nondim-problem-select');
        if (problemSelect) {
            problemSelect.addEventListener('change', updateScalesDisplay);
        }
    }

    // ========================================
    // Global Reset Functions
    // ========================================

    window.resetPendulumExplorer = function() {
        document.getElementById('nondim-mass-slider').value = 1;
        document.getElementById('nondim-length-slider').value = 1;
        document.getElementById('nondim-gravity-slider').value = 9.8;
        updatePendulumPlots();
    };

    window.resetProjectileExplorer = function() {
        document.getElementById('nondim-theta-slider').value = 45;
        document.getElementById('nondim-phi-slider').value = 20;
        document.getElementById('nondim-gamma-slider').value = 0.3;
        updateProjectilePlot();
    };

    // ========================================
    // Auto-initialize
    // ========================================

    initializePlots();

})();
